test/test_iscsi.py

```diff
@@ -0,0 +1,376 @@
+            b'"name":"cinder-zpool/mydataset/volume-f9ecfc53","pool":"poo'\
+            b'l","type":"VOLUME","volsize":{"value":"2G","rawvalue":"2147'\
+            b'483648","parsed":2147483648,"source":"LOCAL"},"used":{"valu'\
+            b'e":"2.06G","rawvalue":"2216689664","parsed":2216689664,"sou'\
+            b'rce":"NONE"},"available":{"value":"50.0G","rawvalue":"53639'\
+            b'102464","parsed":53639102464,"source":"NONE"}}'
+        urlrespond = MagicMock(name="urlrespond")
+        urlrespondcontext = MagicMock(name="urlrespondcontext")
+        urlrespond.__enter__.return_value = urlrespondcontext
+        urlrespondcontext.read.return_value = urlreadresult
+        with patch(request_patch):
+            with patch(open_patch, return_value=urlrespond):
+                self.common.extend_volume(name, new_size)
+
+    def create_export(self, volume_name):
+        return self.common.create_export(volume_name)
+
+
+FakeConnector = {'initiator': 'iqn.2005-10.org.freenas.ctltarget-2b12',
+                 'multipath': False,
+                 'wwpns': ['10000090fa0d6754'],
+                 'wwnns': ['10000090fa0d6755'],
+                 'host': '10.3.1.81',
+                 }
+
+
+FakeSnapshot = {'name': "snap-fakeid", 'volume_name': 'fake-volumeid'}
+
+
+fake_config_dict = {
+    'iscsi_helper': 'tgtadm',
+    'volume_dd_blocksize': 512,
+    'volume_driver':
+        'cinder.volume.drivers.ixsystems.iscsi.FreeNASISCSIDriver',
+    'ixsystems_login': 'root',
+    'ixsystems_password': 'Pa55w0rd',
+    'ixsystems_apikey': '',
+    'ixsystems_server_hostname': '10.3.1.81',
+    'ixsystems_server_port': 80,
+    'ixsystems_transport_type': 'http',
+    'ixsystems_volume_backend_name': 'iXsystems_FREENAS_Storage',
+    'ixsystems_iqn_prefix': 'iqn.2005-10.org.freenas.ctl',
+    'ixsystems_datastore_pool': 'cinder-zpool',
+    'ixsystems_dataset_path': 'cinder-zpool/mydataset',
+    'ixsystems_vendor_name': 'iXsystems',
+    'ixsystems_storage_protocol': 'iscsi',
+    'ixsystems_server_iscsi_port': 3260,
+    'ixsystems_api_version': 'v2.0',
+    'ixsystems_reserved_percentage': 0
+    }
+
+
+class fakeiscsidriver(FreeNASISCSIDriver):
+
+    def __init__(self, configuration):
+        self.configuration = configuration
+        super().__init__(configuration=configuration)
+        self.common = fakecommon(configuration=configuration)
+        self.common.do_custom_setup()
+
+
+@ddt.ddt
+class FreeNASISCSIDriverTestCase(unittest.TestCase):
+
+    def setUp(self):
+        CONF = Mock(spec=conf.Configuration)
+        CONF.iscsi_helper = 'tgtadm'
+        CONF.volume_dd_blocksize = 512
+        CONF.volume_driver = \
+            'cinder.volume.drivers.ixsystems.iscsi.FreeNASISCSIDriver'
+        CONF.ixsystems_login = 'root'
+        CONF.ixsystems_password = 'Pa55w0rd'
+        CONF.ixsystems_apikey = ''
+        CONF.ixsystems_server_hostname = '10.3.1.81'
```

Comment:
Not sure I understand why we are still hardcoding these values? Shouldnt they come from a config file?

Also, where are these used, exactly? hostname and password most importantly? Are we really making HTTP calls or only faking it?

Reply:
All above 4 test scripts test_common.py test_freenasapi.py test_iscsi.py test_utils.py will be uploaded to upstream cinder repository unit test folder as “unit test" . When I check all other existing drivers unit test scripts all of them are fake/mock http calls per spec https://docs.openstack.org/cinder/latest/contributor/testing.html . OpenStack requires the real integration test with real http calls to vendors storage api server in separate integration test system per spec https://wiki.openstack.org/wiki/Cinder/tested-3rdParty-drivers#Driver_Testing , and this is complex and optional for first time driver submission, we might later on provide it separately.  I didn't use external parameter file, where I am trying to follow all others existing unittest cases, none others use parameter file all "hard-coded" into unit test script directly https://github.com/openstack/cinder/tree/master/cinder/tests/unit/volume/drivers and try follow "Mock should be used to remove any external dependencies." state in their requirements spec. I know this is weird, I just hope what we did can pass OpenStack requirement and they can accept ix driver.